types/knex/index.d.ts

```diff
@@ -180,6 +180,7 @@ declare namespace Knex {
     interface Join {
         (raw: Raw): QueryBuilder;
         (tableName: TableName, clause: (this: JoinClause) => void): QueryBuilder;
+        (tableName: TableName, clause: (this: JoinClause, join: JoinClause) => void): QueryBuilder;
```

Comment:
Now you don't need the above overload, just this one.

Reply:
Is that really so? It's still customary (as the tests also indicate) to pass a parameter-less callback to .join() and access the join clause as 'this'. Not having the first overload would not allow that, or am I missing something?

Reply:
In TypeScript you're always allowed to ignore any parameters. So you can write `const x: (s: string) => void = () => {};`.

Reply:
OK, here's a simple join.ts:
```typescript
type JoinClause = string;

function join(this: JoinClause, join: JoinClause) {
    console.log("Hello world");
}

join();
```

Now:
```
$ tsc --noEmit join.ts
join.ts(7,1): error TS2554: Expected 1 arguments, but got 0.
```

Reply:
That's because the function *requires* an argument but you're not *providing* it. That's not allowed, but you can do the opposite: *provide* an argument but not *use* it.

Reply:
Indeed, the tests compile just fine w/o the this-only overload. And thanks for your comments, now I see what I was missing :)